dali/kernels/imgproc/convolution/convolution_gpu.h

```diff
@@ -0,0 +1,215 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_CONVOLUTION_CONVOLUTION_GPU_H_
+#define DALI_KERNELS_IMGPROC_CONVOLUTION_CONVOLUTION_GPU_H_
+
+#include "dali/core/boundary.h"
+#include "dali/core/convert.h"
+#include "dali/core/format.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/imgproc/convolution/cutlass/device/gemm.h"
+#include "dali/kernels/kernel.h"
+#include "dali/pipeline/util/operator_impl_utils.h"
+#include "dali/kernels/imgproc/convolution/cutlass/utility.h"
+
+namespace dali {
+namespace kernels {
+
+/**
+ * @brief Apply convolution with 1-channel `window` in specified axis.
+ *
+ * The window is centered over input and thus it is required to be of odd size.
+ *
+ * Operation is done by using GEMM and generating the matrix from convolution window on the fly
+ *
+ * CUTLASS allows to calculate GEMM: D = alpha * A * B + beta * C. To calculate convolution,
+ * we modify the GEMM to generate one of the input matrices on the fly based on the convolution
+ * window. Such matrix can be generated by applying the convolution window to an indentity matrix
+ * of a desired size.
+ * To the convolution we pass: alpha = `scale`, beta = 0 and D = C = `out`
+ * - we're intersted the multiplication only, the addition is "in-place" (to a zeroed matrix C)
+ * A and B correspond to `in` and `window` (order depends whether it's inner
+ * or outer convolution), effectively calculating D = alpha * A * B.
+ */
+template <typename Out, typename In, typename W, int ndim, int axis, bool has_channels = true>
+struct ConvolutionGpu {
+  KernelRequirements Setup(KernelContext& ctx, const TensorListShape<ndim>& in_shape,
+                           const TensorListShape<1>& window_size) {
+    KernelRequirements req;
+    ScratchpadEstimator se;
+    DALI_ENFORCE(
+        in_shape.size() == window_size.size(),
+        make_string("Provided input shape and window sizes do not mach in number of samples: ",
+                    in_shape.size(), " vs ", window_size.size(), "."));
+    int num_samples = in_shape.size();
+    for (int i = 0; i < num_samples; i++) {
+      int num_channels = has_channels ? in_shape[i][ndim - 1] : 1;
+      DALI_ENFORCE(window_size[i][0] % 2 == 1,
+                   make_string("Kernel window should have odd length, got: ", window_size,
+                               " for sample ", i, "."));
+
+      DALI_ENFORCE((window_size[i][0] / 2) * num_channels < kMaxRadiusSpan,
+                   make_string("Kernel window too big for sample ", i, "."));
```

Comment:
```suggestion
                   make_string("Convolution kernel window is too big for sample ", i, "."));
```
Maybe print window size and maximum size?

Reply:
Either "convolution kernel" or "window", not both.

Reply:
done